lib/plugin_set.py

```diff
@@ -0,0 +1,92 @@
+import os
+import pynsive
+from operator import itemgetter
+from utilities.dict2List import dict2List
+import copy
+
+
+class PluginSet(object):
+    def __init__(self, plugin_location, enabled_plugins=[]):
+        self.plugin_location = plugin_location
+        self.enabled_plugins = self.identify_plugins(enabled_plugins)
+
+    def identify_plugins(self, enabled_plugins):
+        if not os.path.exists(self.plugin_location):
+            return []
+
+        module_name = os.path.basename(self.plugin_location)
+        root_plugin_directory = os.path.join(self.plugin_location, '..')
+
+        plugin_manager = pynsive.PluginManager()
+        plugin_manager.plug_into(root_plugin_directory)
+
+        plugins = []
+
+        found_modules = pynsive.list_modules(module_name)
+        for found_module in found_modules:
+            module_filename, module_name = found_module.split('.')
+            if module_name not in enabled_plugins and len(enabled_plugins) > 0:
+                # Skip this plugin since it's not listed as enabled plugins
+                # or don't skip plugin if we don't have any enabled plugins set (which means all)
+                continue
+
+            module_obj = pynsive.import_module(found_module)
+            reload(module_obj)
+            plugin_class_obj = module_obj.message()
+
+            if 'priority' in dir(plugin_class_obj):
+                priority = plugin_class_obj.priority
+            else:
+                priority = 100
+
+            plugins.append(
+                {
+                    'plugin_class': plugin_class_obj,
+                    'registration': plugin_class_obj.registration,
+                    'priority': priority
+                }
+            )
+        return plugins
+
+    @property
+    def ordered_enabled_plugins(self):
+        return sorted(self.enabled_plugins, key=itemgetter('priority'), reverse=False)
+
+    def run_plugins(self, message, metadata):
+        '''compare the message to the plugin registrations.
+           plugins register with a list of keys or values
+           or values they want to match on
+           this function compares that registration list
+           to the current message and sends the message to plugins
+           in order
+        '''
+        if not isinstance(message, dict):
+            raise TypeError('event is type {0}, should be a dict'.format(type(message)))
+
+        # We deep copy the message in case the caller of this function wants
+        # to reference the original one (so we don't change it in here)
+        message_copy = copy.deepcopy(message)
```

Comment:
This is different, and worth noting. The plugins have the ability to modify a message, so we need to copy the original message so the callers of the plugins, have access to it.

Thoughts?

Reply:
plugins should be able to change a message such that the subsequent plugin gets the new message. This enables things like picking out or correcting an IP in one plugin, geo coding it in another.

Reply:
Plugins still have the ability to do that, there are a couple examples in the tests. This is specifically if a plugin fails, that we won't have a half modified event, and have reference to the original in an esworker for example.

Reply:
@jeffbryner The sendEventToPlugins method this was written based off of, returns the modified message, so this is how callers of the plugins, can handle the message after a plugin has modified it.

Reply:
@jeffbryner I originally misspoke, there isn't a test for it specifically, but subsequent plugins will receive the updated message. I'll circle back though and will add a test for it.

Reply:
OK, as long as each plugin gets the output of the last plugin it sounds like you are doing the same thing as originally. If there is an error caught by the try/catch, where does that get reported so someone know there is a problem?

Reply:
I decided to remove the deepcopy, so the expectation is that both the message that is passed to the function, and the value returned will be the modified message. This allows us to return None in a plugin, and that will indicate that we should drop this message.